File: 110.balanced-binary-tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        try:
            self.height(root)
        except Exception:
            return False
        return True
    
    def height(self, n) -> int:
        if not n:
            return 0
        hl = self.height(n.left)
        hr = self.height(n.right)
        if abs(hl - hr) > 1:
            raise Exception()
        return 1 + max(hl, hr)

# @lc code=end

File: 110.balanced-binary-tree.cpp
/*
 * @lc app=leetcode id=110 lang=cpp
 *
 * [110] Balanced Binary Tree
 */

#include <algorithm>
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode const *root)
    {
        return height(root) >= 0;
    }

private:
    int height(TreeNode const *n)
    {
        if (!n) return 0;
        int hl = height(n->left);
        if (hl < 0) return hl;
        int hr = height(n->right);
        if (hr < 0) return hr;
        if (std::abs(hl - hr) > 1)
            return -1;
        return 1 + std::max(hl, hr);
    }
};
// @lc code=end

